test/performance/odp_ipsecfwd.c

```diff
@@ -0,0 +1,1307 @@
+
+#define MIN(a, b)  (((a) <= (b)) ? (a) : (b))
+
+#define MAX_IFS 2U
+#define MAX_SAS 64U
+#define MAX_FWDS 64U
+#define MAX_SPIS (UINT16_MAX + 1U)
+#define MAX_WORKERS (ODP_THREAD_COUNT_MAX - 1U)
+#define MAX_QUEUES 10U
+#define PKT_SIZE 1024U
+#define PKT_CNT 32768U
+#define WORKER_CNT 1U
+#define NUM_EVS 32U
+
+#define CIPHER_TYPE 0
+#define AUTH_TYPE 1
+#define COMBINED_TYPE 2
+
+#define DIR_IN 0U
+#define DIR_OUT 1U
+
+typedef enum {
+	PRS_OK,
+	PRS_NOK,
+	PRS_TERM
+} parse_result_t;
+
+typedef struct pktio_s pktio_t;
+
+typedef struct pktio_s {
+	union {
+		odp_pktout_queue_t out_dir_qs[MAX_QUEUES];
+		odp_queue_t out_ev_q;
+	};
+
+	odph_ethaddr_t src_mac;
+	char *name;
+	odp_pktio_t handle;
+	int (*send_fn)(const pktio_t *pktio, uint8_t index, odp_packet_t pkt);
+	uint32_t num_tx_qs;
+} pktio_t;
+
+typedef struct {
+	odph_ethaddr_t dst_mac;
+	const pktio_t *pktio;
+	odph_iplookup_prefix_t prefix;
+} fwd_entry_t;
+
+typedef struct {
+	uint64_t ipsec_in_pkts;
+	uint64_t ipsec_out_pkts;
+	uint64_t ipsec_in_errs;
+	uint64_t ipsec_out_errs;
+	uint64_t status_errs;
+	uint64_t fwd_pkts;
+	uint64_t discards;
+} stats_t;
+
+typedef struct prog_config_s prog_config_t;
+
+typedef struct ODP_ALIGNED_CACHE {
+	stats_t stats;
+	prog_config_t *prog_config;
+} thread_config_t;
+
+typedef struct prog_config_s {
+	odph_thread_t thread_tbl[MAX_WORKERS];
+	thread_config_t thread_config[MAX_WORKERS];
+	odp_ipsec_sa_t sas[MAX_SAS];
+	fwd_entry_t fwd_entries[MAX_FWDS];
+	odp_queue_t ev_qs[MAX_SAS];
+	pktio_t pktios[MAX_IFS];
+	char *sa_conf_file;
+	char *fwd_conf_file;
+	odp_instance_t odp_instance;
+	odp_queue_t compl_q;
+	odp_pool_t pktio_pool;
+	odph_table_t fwd_tbl;
+	odp_barrier_t init_barrier;
+	odp_barrier_t term_barrier;
+	uint32_t num_pkts;
+	uint32_t pkt_len;
+	uint32_t num_ifs;
+	uint32_t num_sas;
+	uint32_t num_fwds;
+	int num_thrs;
+	uint8_t mode;
+} prog_config_t;
+
+typedef struct {
+	const char *name;
+	int idx;
+	int type;
+} exposed_alg_t;
+
+static exposed_alg_t exposed_algs[] = {
+	{
+		.idx = ODP_CIPHER_ALG_AES_CBC,
+		.type = CIPHER_TYPE,
+		.name = "ODP_CIPHER_ALG_AES_CBC"
+	},
+	{
+		.idx = ODP_CIPHER_ALG_AES_GCM,
+		.type = COMBINED_TYPE,
+		.name = "ODP_CIPHER_ALG_AES_GCM"
+	},
+	{
+		.idx = ODP_AUTH_ALG_SHA1_HMAC,
+		.type = AUTH_TYPE,
+		.name = "ODP_AUTH_ALG_SHA1_HMAC"
+	}
+};
+
+/* SPIs for in and out directions */
+static odp_ipsec_sa_t *spi_to_sa_map[2U][MAX_SPIS];
+static odp_atomic_u32_t is_running;
+
+static void terminate(int signal ODP_UNUSED)
+{
+	odp_atomic_store_u32(&is_running, 0U);
+}
+
+static void init_config(prog_config_t *config)
+{
+	memset(config, 0, sizeof(*config));
+	config->compl_q = ODP_QUEUE_INVALID;
+	config->pktio_pool = ODP_POOL_INVALID;
+	config->num_thrs = WORKER_CNT;
+}
+
+static void parse_interfaces(prog_config_t *config, const char *optarg)
+{
+	char *tmp_str = strdup(optarg), *tmp = strtok(tmp_str, DELIMITER);
+
+	if (tmp_str == NULL)
+		return;
+
+	if (tmp == NULL)
+		goto out;
```

Comment:
This is redundant with the subsequent while condition. Removing this would get rid of the out label too.